src/browser/history.js

```diff
@@ -0,0 +1,320 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+define((require, exports, module) => {
+  'use strict';
+
+  const PouchDB = require('pouchdb');
+  const {spawn, async} = require('lang/task');
+  const {Record, List, Maybe, Any, Union} = require('typed-immutable/index');
+
+  // PouchDB has a sepcial field `_id` for identifing records
+  // and `_rev` for identifiying revisitions. We will refer to
+  // those properties as `[PouchDB.id]` & `[PouchDB.revision]`
+  // instead.
+  PouchDB.id = "_id";
+  PouchDB.revision = "_rev";
+
+
+  // Type aliases that may be enhanced further sometime later.
+  const ID = String
+  const Revision = String;
+  const TimeStamp = Number;
+  const Type = String;
+  const URI = String;
+  const TagName = String;
+  const Blob = Any;
+
+
+  // Helper function to convert string to a ArrayBuffer instance.
+  const stringToBuffer = string => new TextEncoder().encode(string);
+  // Helper function to create cryptographic hash of the content.
+  const sha = string => crypto.subtle.digest("SHA-256",
+                                             stringToBuffer(string)).then(btoa);
+
+  const stub = record => record.constructor.Stub(record);
+  const hash = record => record.constructor.hash(record);
+
+  const write = (record, db, config) => {
+    return db.put(record.toJSON());
+  };
+
+  const read = async(function*(record, db, config) {
+    try {
+      const data = yield db.get(record[PouchDB.id]);
+      return new record.constructor(data);
+    } catch (error) {
+      if (error.status != 404) {
+        throw error
+      }
+      return record;
+    }
+  });
+
+  const edit = async(function*(record, transform, db, config) {
+    const current = yield read(record, db);
+    const next = yield transform(current);
+    return write(next, db, config);
+  });
+
+  const push = x => xs => xs.push(x);
+  const remove = x => xs => xs.remove(x);
+
+  const exclude = x => xs => {
+    const index = xs.indexOf(x)
+    return index < 0 ? xs :
+           index == 0 ? xs.rest() :
+           index + 1 == xs.size ? xs.butLast() :
+           xs.take(index).concat(xs.skip(index + 1))
+  }
+  const include = x => xs => xs.indexOf(x) < 0 ? xs.push(x) : xs
+
+
+  /*
+  ## Interests API
+
+  ### Sites
```

Comment:
This should probably be called `pages` or `locations`, not `sites`. Sites are root domain, like `facebook.com`. I think we will want to surface only root domains in some places like favorites tiles, so it would be better if we got the terminology right up front.